File: backend/agents/types/LLMCouncil.ts
/**
 * LLM Council Multi-Model Decision Making Types
 *
 * Week 52: Supports 3-stage consensus process
 * - Stage 1: Response (query all models in parallel)
 * - Stage 2: Peer Review (blind review of responses)
 * - Stage 3: Synthesis (chairman creates final decision)
 */

// ============================================================================
// ENUMS
// ============================================================================

export enum SessionStatus {
    PENDING = 'pending',
    REVIEWING = 'reviewing',
    COMPLETE = 'complete'
}

export enum DecisionType {
    ARCHITECTURE = 'architecture',
    PLANNING = 'planning',
    QUALITY = 'quality',
    GENERATION = 'generation'
}

// ============================================================================
// CORE INTERFACES
// ============================================================================

export interface CouncilSession {
    id: string;
    question: string;
    context: Record<string, any>;
    decision_type: DecisionType;
    agent_id: string;
    status: SessionStatus;
    created_at: Date;
    completed_at?: Date;
    // Relationships (populated via eager loading)
    responses?: CouncilResponse[];
    reviews?: CouncilReview[];
    decision?: CouncilDecision;
}

export interface CouncilResponse {
    id: string;
    session_id: string;
    model_name: string;
    response_text: string;
    confidence: number;  // 0.0-1.0
    reasoning?: string;
    created_at: Date;
    // Relationships
    reviews_received?: CouncilReview[];
}

export interface CouncilReview {
    id: string;
    session_id: string;
    reviewer_model: string;
    reviewed_response_id: string;
    accuracy_score: number;      // 0-10
    completeness_score: number;  // 0-10
    clarity_score: number;       // 0-10
    feasibility_score: number;   // 0-10
    comments?: string;
    created_at: Date;
}

export interface CouncilDecision {
    id: string;
    session_id: string;
    chairman_model: string;
    final_decision: string;
    confidence: number;  // 0.0-1.0
    consensus_level: number;  // 0-100% agreement
    dissenting_opinions?: string[];
    synthesis_reasoning?: string;
    created_at: Date;
}

// ============================================================================
// API REQUEST/RESPONSE INTERFACES
// ============================================================================

export interface CreateSessionRequest {
    question: string;
    context: Record<string, any>;
    decision_type: DecisionType;
    agent_id: string;
}

export interface CreateSessionResponse {
    session: CouncilSession;
    message: string;
}

export interface QuickDecisionRequest {
    question: string;
    context: Record<string, any>;
    decision_type: DecisionType;
    agent_id: string;
    wait_for_completion?: boolean;  // Default: true
}

export interface QuickDecisionResponse {
    session: CouncilSession;
    decision?: CouncilDecision;  // Only if wait_for_completion=true
    message: string;
}

export interface SessionListResponse {
    sessions: CouncilSession[];
    total: number;
    page: number;
    page_size: number;
}

export interface SessionDetailResponse {
    session: CouncilSession;
    responses: CouncilResponse[];
    reviews: CouncilReview[];
    decision?: CouncilDecision;
    consensus_metrics: ConsensusMetrics;
}

export interface ConsensusMetrics {
    total_responses: number;
    average_confidence: number;
    confidence_std_dev: number;
    consensus_level: number;  // 0-100%
    has_outliers: boolean;
    outlier_responses?: string[];  // response IDs
    review_scores: {
        average_accuracy: number;
        average_completeness: number;
        average_clarity: number;
        average_feasibility: number;
        overall_average: number;
    };
}

// ============================================================================
// CONFIGURATION INTERFACES
// ============================================================================

export interface CouncilModelConfig {
    model: string;
    weight: number;
    role: string;
    specialty?: string;
}

export interface CouncilModelsConfig {
    chairman: {
        model: string;  // "deepseek-r1:latest"
        weight: number; // 2.0
        role: string;   // "chairman"
    };
    council: CouncilModelConfig[];
}

export const DEFAULT_MODELS_CONFIG: CouncilModelsConfig = {
    chairman: {
        model: 'deepseek-r1:latest',
        weight: 2.0,
        role: 'chairman'
    },
    council: [
        {
            model: 'qwen2.5-coder:7b',
            weight: 1.5,
            role: 'technical',
            specialty: 'code architecture & design patterns'
        },
        {
            model: 'codellama:latest',
            weight: 1.5,
            role: 'implementation',
            specialty: 'code implementation & debugging'
        },
        {
            model: 'mistral:latest',
            weight: 1.0,
            role: 'documentation',
            specialty: 'technical writing & clarity'
        },
        {
            model: 'qwen2.5:7b',
            weight: 1.0,
            role: 'planning',
            specialty: 'project planning & estimation'
        },
        {
            model: 'llama3.2:latest',
            weight: 1.0,
            role: 'quality',
            specialty: 'quality assurance & testing'
        }
    ]
};

// ============================================================================
// HELPER FUNCTIONS
// ============================================================================

/**
 * Calculate consensus level from responses
 *
 * Uses weighted voting based on model weights and confidence scores.
 * Returns percentage (0-100) of agreement among models.
 */
export function calculateConsensus(
    responses: CouncilResponse[],
    modelsConfig: CouncilModelsConfig
): number {
    if (responses.length === 0) return 0;

    // Extract key recommendations from each response
    const recommendations = responses.map(r =>
        extractRecommendation(r.response_text)
    );

    // Calculate similarity between all pairs
    let totalWeight = 0;
    let agreementWeight = 0;

    for (let i = 0; i < responses.length; i++) {
        for (let j = i + 1; j < responses.length; j++) {
            const model1 = responses[i].model_name;
            const model2 = responses[j].model_name;

            const weight1 = getModelWeight(model1, modelsConfig);
            const weight2 = getModelWeight(model2, modelsConfig);
            const pairWeight = (weight1 + weight2) / 2;

            totalWeight += pairWeight;

            // Check if recommendations are similar
            const similarity = calculateTextSimilarity(
                recommendations[i],
                recommendations[j]
            );

            if (similarity > 0.7) {  // 70% similarity threshold
                agreementWeight += pairWeight;
            }
        }
    }

    return totalWeight > 0 ? (agreementWeight / totalWeight) * 100 : 0;
}

/**
 * Detect outlier responses that significantly differ from consensus
 *
 * Returns IDs of responses that are >2 standard deviations from mean confidence
 * or have recommendation text that differs significantly from others.
 */
export function detectOutliers(
    responses: CouncilResponse[]
): string[] {
    if (responses.length < 3) return [];  // Need at least 3 for outlier detection

    const outliers: string[] = [];

    // 1. Check confidence outliers
    const confidences = responses.map(r => r.confidence);
    const meanConfidence = confidences.reduce((a, b) => a + b, 0) / confidences.length;
    const stdDev = Math.sqrt(
        confidences.reduce((sum, c) => sum + Math.pow(c - meanConfidence, 2), 0) / confidences.length
    );

    responses.forEach(r => {
        if (Math.abs(r.confidence - meanConfidence) > 2 * stdDev) {
            if (!outliers.includes(r.id)) {
                outliers.push(r.id);
            }
        }
    });

    // 2. Check recommendation text outliers
    const recommendations = responses.map(r =>
        extractRecommendation(r.response_text)
    );

    responses.forEach((r, idx) => {
        let totalSimilarity = 0;
        let comparisons = 0;

        recommendations.forEach((otherRec, otherIdx) => {
            if (idx !== otherIdx) {
                totalSimilarity += calculateTextSimilarity(
                    recommendations[idx],
                    otherRec
                );
                comparisons++;
            }
        });

        const avgSimilarity = totalSimilarity / comparisons;
        if (avgSimilarity < 0.3) {  // Less than 30% similar to others
            if (!outliers.includes(r.id)) {
                outliers.push(r.id);
            }
        }
    });

    return outliers;
}

/**
 * Calculate average review scores for a response
 */
export function calculateAverageReviewScores(
    reviews: CouncilReview[]
): {
    accuracy: number;
    completeness: number;
    clarity: number;
    feasibility: number;
    overall: number;
} {
    if (reviews.length === 0) {
        return { accuracy: 0, completeness: 0, clarity: 0, feasibility: 0, overall: 0 };
    }

    const sum = reviews.reduce((acc, r) => ({
        accuracy: acc.accuracy + r.accuracy_score,
        completeness: acc.completeness + r.completeness_score,
        clarity: acc.clarity + r.clarity_score,
        feasibility: acc.feasibility + r.feasibility_score
    }), { accuracy: 0, completeness: 0, clarity: 0, feasibility: 0 });

    const count = reviews.length;
    const accuracy = sum.accuracy / count;
    const completeness = sum.completeness / count;
    const clarity = sum.clarity / count;
    const feasibility = sum.feasibility / count;
    const overall = (accuracy + completeness + clarity + feasibility) / 4;

    return { accuracy, completeness, clarity, feasibility, overall };
}

// ============================================================================
// PRIVATE HELPER FUNCTIONS
// ============================================================================

function extractRecommendation(responseText: string): string {
    // Extract text after "RECOMMENDATION:" marker
    const match = responseText.match(/RECOMMENDATION:\s*(.+?)(?=\n|REASONING:|CONFIDENCE:|$)/s);
    return match ? match[1].trim() : responseText.substring(0, 200);
}

function getModelWeight(
    modelName: string,
    config: CouncilModelsConfig
): number {
    if (config.chairman.model === modelName) {
        return config.chairman.weight;
    }

    const councilModel = config.council.find(m => m.model === modelName);
    return councilModel ? councilModel.weight : 1.0;
}

function calculateTextSimilarity(text1: string, text2: string): number {
    // Simple Jaccard similarity based on word sets
    const words1 = new Set(text1.toLowerCase().split(/\s+/));
    const words2 = new Set(text2.toLowerCase().split(/\s+/));

    const intersection = new Set([...words1].filter(w => words2.has(w)));
    const union = new Set([...words1, ...words2]);

    return intersection.size / union.size;
}

File: backend/app/models/llm_council.py
"""
LLM Council Models for Multi-Model Decision Making

Week 52: LLM Council Multi-Model Decision Making
Supports 3-stage consensus process: Response → Peer Review → Synthesis
"""

from sqlalchemy import Column, String, Text, Float, DateTime, ForeignKey, CheckConstraint, Index, ARRAY
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
import uuid

from app.database import Base


class CouncilSession(Base):
    """
    Council session for multi-LLM decision making.

    A session represents one question/decision that needs council input.
    Goes through 3 stages: Response → Review → Synthesis
    """
    __tablename__ = "council_sessions"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    question = Column(Text, nullable=False)
    context = Column(JSONB, nullable=False, default={})
    decision_type = Column(String(50), nullable=False)  # architecture, planning, quality, generation
    agent_id = Column(String(50), nullable=False, index=True)
    status = Column(String(20), nullable=False, default="pending")  # pending, reviewing, complete
    created_at = Column(DateTime, nullable=False, server_default=func.now())
    completed_at = Column(DateTime, nullable=True)

    # Relationships
    responses = relationship(
        "CouncilResponse",
        back_populates="session",
        cascade="all, delete-orphan",
        lazy="selectin"
    )
    reviews = relationship(
        "CouncilReview",
        back_populates="session",
        cascade="all, delete-orphan",
        lazy="selectin"
    )
    decision = relationship(
        "CouncilDecision",
        back_populates="session",
        uselist=False,
        cascade="all, delete-orphan",
        lazy="selectin"
    )

    # Table constraints
    __table_args__ = (
        CheckConstraint(
            status.in_(['pending', 'reviewing', 'complete']),
            name='ck_council_session_status'
        ),
        CheckConstraint(
            decision_type.in_(['architecture', 'planning', 'quality', 'generation']),
            name='ck_council_session_decision_type'
        ),
        Index('idx_council_session_agent_status', 'agent_id', 'status'),
    )

    def __repr__(self):
        return f"<CouncilSession(id={self.id}, type={self.decision_type}, status={self.status})>"


class CouncilResponse(Base):
    """
    Individual model response in a council session.

    Each LLM model provides its own response to the question.
    Stage 1 of the 3-stage process.
    """
    __tablename__ = "council_responses"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    session_id = Column(
        UUID(as_uuid=True),
        ForeignKey("council_sessions.id", ondelete="CASCADE"),
        nullable=False,
        index=True
    )
    model_name = Column(String(100), nullable=False)
    response_text = Column(Text, nullable=False)
    confidence = Column(Float, nullable=False)  # 0.0-1.0
    reasoning = Column(Text, nullable=True)
    created_at = Column(DateTime, nullable=False, server_default=func.now())

    # Relationships
    session = relationship("CouncilSession", back_populates="responses")
    reviews_received = relationship(
        "CouncilReview",
        back_populates="reviewed_response",
        cascade="all, delete-orphan",
        lazy="selectin"
    )

    # Table constraints
    __table_args__ = (
        CheckConstraint(
            confidence.between(0.0, 1.0),
            name='ck_council_response_confidence'
        ),
        Index('idx_council_response_session', 'session_id'),
    )

    def __repr__(self):
        return f"<CouncilResponse(id={self.id}, model={self.model_name}, confidence={self.confidence})>"


class CouncilReview(Base):
    """
    Peer review from one model to another's response.

    Anonymous/blind review where models evaluate each other's responses.
    Stage 2 of the 3-stage process.
    """
    __tablename__ = "council_reviews"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    session_id = Column(
        UUID(as_uuid=True),
        ForeignKey("council_sessions.id", ondelete="CASCADE"),
        nullable=False,
        index=True
    )
    reviewer_model = Column(String(100), nullable=False)
    reviewed_response_id = Column(
        UUID(as_uuid=True),
        ForeignKey("council_responses.id", ondelete="CASCADE"),
        nullable=False
    )
    accuracy_score = Column(Float, nullable=False)      # 0-10
    completeness_score = Column(Float, nullable=False)  # 0-10
    clarity_score = Column(Float, nullable=False)       # 0-10
    feasibility_score = Column(Float, nullable=False)   # 0-10
    comments = Column(Text, nullable=True)
    created_at = Column(DateTime, nullable=False, server_default=func.now())

    # Relationships
    session = relationship("CouncilSession", back_populates="reviews")
    reviewed_response = relationship("CouncilResponse", back_populates="reviews_received")

    # Table constraints
    __table_args__ = (
        CheckConstraint(
            accuracy_score.between(0.0, 10.0),
            name='ck_council_review_accuracy'
        ),
        CheckConstraint(
            completeness_score.between(0.0, 10.0),
            name='ck_council_review_completeness'
        ),
        CheckConstraint(
            clarity_score.between(0.0, 10.0),
            name='ck_council_review_clarity'
        ),
        CheckConstraint(
            feasibility_score.between(0.0, 10.0),
            name='ck_council_review_feasibility'
        ),
        Index('idx_council_review_session', 'session_id'),
        Index('idx_council_review_response', 'reviewed_response_id'),
    )

    def __repr__(self):
        return f"<CouncilReview(id={self.id}, reviewer={self.reviewer_model})>"

    @property
    def average_score(self) -> float:
        """Calculate average of all review scores."""
        return (
            self.accuracy_score +
            self.completeness_score +
            self.clarity_score +
            self.feasibility_score
        ) / 4.0


class CouncilDecision(Base):
    """
    Final synthesized decision from chairman model.

    The chairman (deepseek-r1) synthesizes all responses and reviews
    into a single consensus decision.
    Stage 3 of the 3-stage process.
    """
    __tablename__ = "council_decisions"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    session_id = Column(
        UUID(as_uuid=True),
        ForeignKey("council_sessions.id", ondelete="CASCADE"),
        nullable=False,
        unique=True,
        index=True
    )
    chairman_model = Column(String(100), nullable=False)
    final_decision = Column(Text, nullable=False)
    confidence = Column(Float, nullable=False)  # 0.0-1.0
    consensus_level = Column(Float, nullable=False)  # 0-100% agreement
    dissenting_opinions = Column(ARRAY(Text), nullable=True)  # Minority views
    synthesis_reasoning = Column(Text, nullable=True)  # Why this decision?
    created_at = Column(DateTime, nullable=False, server_default=func.now())

    # Relationships
    session = relationship("CouncilSession", back_populates="decision")

    # Table constraints
    __table_args__ = (
        CheckConstraint(
            confidence.between(0.0, 1.0),
            name='ck_council_decision_confidence'
        ),
        CheckConstraint(
            consensus_level.between(0.0, 100.0),
            name='ck_council_decision_consensus'
        ),
        Index('idx_council_decision_session', 'session_id'),
    )

    def __repr__(self):
        return f"<CouncilDecision(id={self.id}, consensus={self.consensus_level}%)>"

    @property
    def has_strong_consensus(self) -> bool:
        """Check if consensus level is strong (>80%)."""
        return self.consensus_level >= 80.0

    @property
    def has_dissent(self) -> bool:
        """Check if there are dissenting opinions."""
        return bool(self.dissenting_opinions)
